### PR TITLE
move the code from runtime.OptIn*, runtime.addApp(..)

### PR DESCRIPTION
**Pull request by amityadav0**
- reuse functions
- fix transaction loading

### CODE REVIEW

File: examples/permissioned-token-freezing/test/asset-txfer-test.js [hunk 1]
```diff
@@ -46,6 +46,8 @@ describe('Test for transferring asset using custom logic', function () {
     assert.equal(assetDef.clawback, alice.address);
 
     runtime.optIntoASA(assetId, bob.address, {});
+    escrow = undefined;
```
Comment (ratik21) [inline examples/permissioned-token-freezing/test/asset-txfer-test.js:49]: why do we need this? In `syncAccounts` we do 
```js
if (escrow) { escrow = runtime.getAccount(escrow.address); }
```

File: packages/runtime/src/ctx.ts [hunk 2]
```diff
@@ -8,10 +8,10 @@ import {
   AccountAddress, AccountStoreI, AlgoTransferParam, ASADeploymentFlags, AssetHoldingM, AssetModFields,
   AssetTransferParam, Context, ExecParams, ExecutionMode,
   SignType, SSCAttributesM, SSCDeploymentFlags, SSCOptionalFlags,
-  State, TransactionType, Txn, TxParams
+  State, TransactionType, Txn, TxParams, UpdateSSCParam
 } from "./types";
 
-const approvalProgram = "approval-program";
+const approvalProgramConst = "approval-program";
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:14]: `const approvalProgramConst` doesn't seem right. Maybe we could rename to `const APPROVAL_PROGRAM`?

File: packages/runtime/src/ctx.ts [hunk 3]
```diff
@@ -133,6 +133,46 @@ export class Ctx implements Context {
     }
   }
 
+  /**
+   * Add Asset
+   * @param name ASA name
+   * @param flags ASA Deployment Flags
+   */
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:140]: Please update the description

File: packages/runtime/src/ctx.ts [hunk 4]
```diff
@@ -133,6 +133,46 @@ export class Ctx implements Context {
...
+      this.runtime.loadedAssetsDefs[name]
+    );
+    this.runtime.mkAssetCreateTx(name, flags, asset);
+
+    this.state.assetDefs.set(this.state.assetCounter, senderAcc.address);
+    this.state.assetNameInfo.set(name, {
+      creator: senderAcc.address,
+      assetIndex: this.state.assetCounter,
+      assetDef: asset,
+      txId: "tx-id",
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:156]: Could we update to `this.tx.txID()`?

File: packages/runtime/src/ctx.ts [hunk 5]
```diff
@@ -133,6 +133,46 @@ export class Ctx implements Context {
...
+      firstValid: flags.firstValid,
+      validRounds: flags.validRounds,
+      lease: flags.lease,
+      note: flags.note,
+      noteb64: flags.noteb64,
+      closeRemainderTo: flags.closeRemainderTo,
+      rekeyTo: flags.rekeyTo
+    };
+    this.optIntoASA(this.state.assetCounter, senderAcc.address, payFlags); // opt-in for creator
+
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:172]:
```suggestion
```

File: packages/runtime/src/ctx.ts [hunk 6]
```diff
@@ -157,6 +197,71 @@ export class Ctx implements Context {
     account.optInToASA(assetIndex, assetHolding);
   }
 
+  /**
+   * creates new application and returns application id
+   * @param flags SSCDeployment flags
+   * @param payFlags Transaction parameters
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:203]: please update these params

File: packages/runtime/src/ctx.ts [hunk 7]
```diff
@@ -157,6 +197,71 @@ export class Ctx implements Context {
...
+    const attributes = this.getApp(0);
+    senderAcc.createdApps.delete(0); // remove zero app from sender's account
+    this.state.globalApps.delete(0); // remove zero app from context
+    senderAcc.createdApps.set(this.state.appCounter, attributes);
+    this.state.appNameInfo.set(
+      approvalProgram + "-" + clearProgram,
+      {
+        creator: senderAcc.address,
+        appID: this.state.appCounter,
+        txId: "tx-id",
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:240]: I think we can use `this.tx.txID()`

File: packages/runtime/src/ctx.ts [hunk 8]
```diff
@@ -157,6 +197,71 @@ export class Ctx implements Context {
...
+   * @param appId Application Id
+   * @param flags Stateful smart contract transaction optional parameters (accounts, args..)
+   * @param payFlags Transaction Parameters
+   */
+  optInToApp (accountAddr: string, appID: number): void {
+    const appParams = this.getApp(appID);
+
+    const account = this.getAccount(accountAddr);
+    account.optInToApp(appID, appParams);
+
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:261]:
```suggestion
```

File: packages/runtime/src/ctx.ts [hunk 9]
```diff
@@ -157,6 +197,71 @@ export class Ctx implements Context {
...
+    return this.state.appCounter;
+  }
+
+  /**
+   * Account address opt-in for application Id
+   * @param accountAddr Account address
+   * @param appId Application Id
+   * @param flags Stateful smart contract transaction optional parameters (accounts, args..)
+   * @param payFlags Transaction Parameters
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:254]: description params are not correct

File: packages/runtime/src/ctx.ts [hunk 10]
```diff
@@ -307,14 +412,32 @@ export class Ctx implements Context {
...
   /**
-   * Update Apps
-   * @param appID  Application index
+   * Update application
+   * @param senderAddr sender address
+   * @param appId application Id
    * @param approvalProgram new approval program
    * @param clearProgram new clear program
+   * @param payFlags Transaction parameters
+   * @param flags Stateful smart contract transaction optional parameters (accounts, args..)
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:421]: Please update the params

File: packages/runtime/src/ctx.ts [hunk 11]
```diff
@@ -430,94 +554,43 @@ export class Ctx implements Context {
...
-            accounts: txnParam.accounts,
-            foreignApps: txnParam.foreignApps,
-            foreignAssets: txnParam.foreignAssets,
-            note: txnParam.note,
-            lease: txnParam.lease
-          };
-          this.runtime.addCtxOptInTx(fromAccountAddr, txnParam.appID, txnParam.payFlags, flags);
-          const account = this.getAccount(fromAccountAddr);
-          account.optInToApp(txnParam.appID, appParams);
+          this.tx = this.gtxs[idx]; // update current tx to the requested index
```
Comment (ratik21) [inline packages/runtime/src/ctx.ts:591]:
```suggestion
          this.tx = this.gtxs[idx]; // update current tx to txn being exectuted in group
```

File: packages/runtime/src/types.ts [hunk 12]
```diff
@@ -121,6 +121,14 @@ export interface Context {
   deleteApp: (appId: number) => void
   closeApp: (sender: AccountAddress, appId: number) => void
   processTransactions: (txnParams: ExecParams[]) => void
+  addAsset: (name: string, fromAccountAddr: string, flags: ASADeploymentFlags) => number
```
Comment (ratik21) [inline packages/runtime/src/types.ts:124]:
```suggestion
  addAsset: (name: string, fromAccountAddr: AccountAddress, flags: ASADeploymentFlags) => number
```

**Review by ratik21 — APPROVED** [review]
Looks good. Left few suggestions & comments

File: examples/permissioned-token-freezing/test/asset-txfer-test.js (hunk 1 above)
Comment (amityadav0) [reply examples/permissioned-token-freezing/test/asset-txfer-test.js:49]: escrow has some address loaded somehow.

Comment (amityadav0) [reply examples/permissioned-token-freezing/test/asset-txfer-test.js:49]: probably when it moves to second test, global value escrow is updated from previous test.

File: packages/runtime/src/ctx.ts (hunk 2 above)
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:14]: yes

File: packages/runtime/src/ctx.ts (hunk 3 above)
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:140]: ok

File: packages/runtime/src/ctx.ts (hunk 4 above)
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:156]: to `this.tx.txID` we can

**Pull request merged**